File: internal/server/auth/memories.go
package auth

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"
)

// Memory CRUD routes (Stage 1 of plan/rag-and-memory-roadmap.md).
//
// All routes require a logged-in user. Authorization is *not* shared across
// users: the WHERE clauses in store_memories.go always include user_id, so
// even a forged memory ID can't reach another user's row.
//
// Route layout:
//
//   GET    /api/auth/memories                — list all of the caller's memories
//   POST   /api/auth/memories                — create
//   PUT    /api/auth/memories/{id}           — update content
//   DELETE /api/auth/memories/{id}           — delete
//
// Per-prompt fetch (the agent path) calls Store.ListUserMemoriesForPrompt
// directly — it doesn't hit HTTP at all.

const memoryMaxContentBytes = 2 * 1024 // 2KB per memory — keeps the prompt budget predictable

func writeMemoryError(w http.ResponseWriter, status int, code string) {
	writeJSON(w, status, map[string]any{"error": code})
}

type memoryRequest struct {
	WorkspaceID string `json:"workspaceId,omitempty"`
	Content     string `json:"content"`
}

func (h *Handler) handleMemories(w http.ResponseWriter, r *http.Request) {
	user, _, err := h.Authenticate(r)
	if err != nil {
		writeMemoryError(w,http.StatusUnauthorized, "unauthorized")
		return
	}
	switch r.Method {
	case http.MethodGet:
		ms, err := h.Store.ListAllUserMemories(r.Context(), user.ID)
		if err != nil {
			writeMemoryError(w,http.StatusInternalServerError, "list_failed")
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"memories": ms})
	case http.MethodPost:
		var req memoryRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeMemoryError(w,http.StatusBadRequest, "invalid_json")
			return
		}
		content := strings.TrimSpace(req.Content)
		if content == "" {
			writeMemoryError(w,http.StatusBadRequest, "empty_content")
			return
		}
		if len(content) > memoryMaxContentBytes {
			writeMemoryError(w,http.StatusBadRequest, "content_too_long")
			return
		}
		m, err := h.Store.CreateUserMemory(r.Context(), user.ID, strings.TrimSpace(req.WorkspaceID), content)
		if err != nil {
			writeMemoryError(w,http.StatusInternalServerError, "create_failed")
			return
		}
		writeJSON(w, http.StatusOK, m)
	default:
		w.Header().Set("Allow", "GET, POST")
		writeMemoryError(w,http.StatusMethodNotAllowed, "method_not_allowed")
	}
}

func (h *Handler) handleMemoriesSubtree(w http.ResponseWriter, r *http.Request) {
	user, _, err := h.Authenticate(r)
	if err != nil {
		writeMemoryError(w,http.StatusUnauthorized, "unauthorized")
		return
	}
	idStr := strings.TrimPrefix(r.URL.Path, "/api/auth/memories/")
	idStr = strings.TrimSuffix(idStr, "/")
	if idStr == "" || strings.Contains(idStr, "/") {
		http.NotFound(w, r)
		return
	}
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		writeMemoryError(w,http.StatusBadRequest, "invalid_id")
		return
	}
	switch r.Method {
	case http.MethodPut:
		var req memoryRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeMemoryError(w,http.StatusBadRequest, "invalid_json")
			return
		}
		content := strings.TrimSpace(req.Content)
		if content == "" {
			writeMemoryError(w,http.StatusBadRequest, "empty_content")
			return
		}
		if len(content) > memoryMaxContentBytes {
			writeMemoryError(w,http.StatusBadRequest, "content_too_long")
			return
		}
		m, err := h.Store.UpdateUserMemory(r.Context(), user.ID, id, content)
		if err != nil {
			if errors.Is(err, ErrMemoryNotFound) {
				writeMemoryError(w,http.StatusNotFound, "not_found")
				return
			}
			writeMemoryError(w,http.StatusInternalServerError, "update_failed")
			return
		}
		writeJSON(w, http.StatusOK, m)
	case http.MethodDelete:
		if err := h.Store.DeleteUserMemory(r.Context(), user.ID, id); err != nil {
			if errors.Is(err, ErrMemoryNotFound) {
				writeMemoryError(w,http.StatusNotFound, "not_found")
				return
			}
			writeMemoryError(w,http.StatusInternalServerError, "delete_failed")
			return
		}
		w.WriteHeader(http.StatusNoContent)
	default:
		w.Header().Set("Allow", "PUT, DELETE")
		writeMemoryError(w,http.StatusMethodNotAllowed, "method_not_allowed")
	}
}

File: frontend/src/lib/memories.ts
// memories — Stage 1 RAG roadmap. Wraps /api/auth/memories CRUD.
//
// 메모리는 사용자별로 묶이고, workspaceId 가 비어 있으면 전역 (사용자가
// 들어가는 모든 워크스페이스), 있으면 그 워크스페이스에서만 LLM 프롬프트에
// inject 된다. 우선순위 (시스템 > 사용자 메모리 > 태스크 지시) 는 서버
// agent.go 가 강제하므로 클라이언트는 CRUD 만 책임진다.

export type UserMemory = {
  id: number;
  userId: string;
  workspaceId?: string;
  content: string;
  createdAt: number;
  updatedAt: number;
};

export const MEMORY_MAX_CONTENT_LENGTH = 2048;

async function readJSON<T>(res: Response): Promise<T> {
  const ct = res.headers.get("content-type") ?? "";
  if (!ct.includes("application/json")) {
    throw new Error(`unexpected content-type: ${ct}`);
  }
  return (await res.json()) as T;
}

function codeOrStatus(res: Response, body: unknown): string {
  if (body && typeof body === "object" && "error" in body) {
    const e = (body as { error?: unknown }).error;
    if (typeof e === "string") return e;
  }
  return `http_${res.status}`;
}

export async function fetchMemories(): Promise<UserMemory[]> {
  const res = await fetch("/api/auth/memories", { credentials: "include" });
  if (!res.ok) throw new Error(`memories list failed: ${res.status}`);
  const body = await readJSON<{ memories: UserMemory[] }>(res);
  return body.memories ?? [];
}

export async function createMemory(content: string, workspaceId?: string): Promise<UserMemory> {
  const res = await fetch("/api/auth/memories", {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content, workspaceId: workspaceId ?? "" }),
  });
  if (!res.ok) {
    let body: unknown = null;
    try { body = await res.json(); } catch { /* ignore */ }
    throw new Error(codeOrStatus(res, body));
  }
  return readJSON<UserMemory>(res);
}

export async function updateMemory(id: number, content: string): Promise<UserMemory> {
  const res = await fetch(`/api/auth/memories/${id}`, {
    method: "PUT",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
  if (!res.ok) {
    let body: unknown = null;
    try { body = await res.json(); } catch { /* ignore */ }
    throw new Error(codeOrStatus(res, body));
  }
  return readJSON<UserMemory>(res);
}

export async function deleteMemory(id: number): Promise<void> {
  const res = await fetch(`/api/auth/memories/${id}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!res.ok && res.status !== 204) {
    throw new Error(`memory delete failed: ${res.status}`);
  }
}
